### PR TITLE
[SPARK-38477][CORE] Use error class in org.apache.spark.shuffle

### PR DESCRIPTION
**Pull request by bozhang2820**
### What changes were proposed in this pull request?
This PR aims to change exceptions created in package org.apache.spark.shuffle to use error class. This also adds an error class INTERNAL_ERROR_SHUFFLE and uses that for the internal errors in the package.

### Why are the changes needed?
This is to move exceptions created in package org.apache.spark.shuffle to error class.

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
Existing tests.

### CODE REVIEW

File: core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala:267]: @bozhang2820 Are you sure that IO errors should be considered as internal errors? Users can face to them in real life, I guess.

Comment (bozhang2820) [reply core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala:267]: Sorry for the late reply. Yes I think we might need a `SparkIOException` which extends `IOException`?

Comment (MaxGekk) [reply core/src/main/scala/org/apache/spark/shuffle/IndexShuffleBlockResolver.scala:267]: I think that re-using the existing one `SparkRuntimeException` with a proper error class should be enough but `SparkIOException` works to me as well.

File: core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala
Comment (MaxGekk) [inline core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:472]: Can't you re-using the existing one `FAILED_RENAME_PATH`?

Comment (bozhang2820) [reply core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:472]: `FAILED_RENAME_PATH` is for the case when the destination already exists. `_LEGACY_ERROR_TEMP_2199` suits our case better.

Comment (MaxGekk) [reply core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala:472]: Can you assign a proper name?

Comment (bozhang2820) [reply core/src/main/scala/org/apache/spark/errors/SparkCoreErrors.scala]: Assigned a name to the error class.

**Comment by MaxGekk** [comment]
@Hisoka-X Could you make the assert message shorter, and print only `Error class: INTERNAL_ERROR_SHUFFLE is not in sql-error-conditions.md`, see https://github.com/bozhang2820/spark/actions/runs/5465673291/jobs/9949470464

File: common/utils/src/main/resources/error/error-classes.json
Comment (MaxGekk) [inline common/utils/src/main/resources/error/error-classes.json:729]: Which rename returned false. Could you make it clearer? Do you mean rename of filesystem.

**Review by MaxGekk — COMMENTED** [review]
@bozhang2820 Please, resolve conflicts and fix a test failure.

File: common/utils/src/main/resources/error/error-classes.json
Comment (bozhang2820) [reply common/utils/src/main/resources/error/error-classes.json:729]: Change to `FileSystem.rename`.

**Review by MaxGekk — APPROVED** [review]
Waiting for CI.

**Comment by MaxGekk** [comment]
@bozhang2820 Could you resolve conflicts, please.

**Review by holdenk — APPROVED** [review]
LGTM pending @MaxGekk

**Comment by MaxGekk** [comment]
+1, LGTM. Merging to master.
Thank you, @bozhang2820 and @holdenk for review.

**Pull request closed without merging**
